src/shogun/mathematics/linalg/LinalgNamespace.h

```diff
@@ -102,118 +102,162 @@ LinalgBackendBase* infer_backend(const Container<T>& a, const Container<T>& b)
 }
 
 /**
- * Transfers data to GPU memory. Does nothing if no GPU backend registered.
+ * Transfers data to GPU memory.
+ * Shallow-copy of SGVector with vector on CPU if GPU backend not available
  *
- * @param vector SGVector to be transferred
- * @return SGVector with vector on GPU if GPU backend is available
- * and a shallow-copy of SGVector with vector on CPU if GPU backend not available
+ * @param a SGVector to be transferred
+ * @param b SGVector to be set
  */
 template <typename T>
-SGVector<T> to_gpu(const SGVector<T>& vector)
+void to_gpu(SGVector<T>& a, SGVector<T>& b)
 {
-	REQUIRE(!vector.on_gpu(), "The vector is already on GPU.\n");
-	LinalgBackendBase* gpu_backend = sg_linalg->get_gpu_backend();
-	if (gpu_backend)
-		return SGVector<T>(gpu_backend->to_gpu(vector), vector.vlen);
+	sg_linalg->m_gpu_transfer.lock();
```

Comment:
I don't think the lock should be in the else branch. For example, assuming `a` is a GPU vector, thread1 is transferring gpu vector `a` from GPU back to CPU, and during which time `a.on_gpu() == true`. If thread2 calls `to_gpu()` when thread1 is working, without the lock in the beginning, the `if` clause will be `true` and nothing will be done. Thus, after thread1 and thread2 (`from_gpu()` and `to_gpu()`), `a` will end up in CPU. However with the lock, `a` will be transferred to CPU and moved back to GPU.